File: 0239-sliding-window-maximum/0239-sliding-window-maximum.py
from collections import deque
class Solution:
    def maxSlidingWindow(self, nums, k):
        if not nums or k <= 0:
            return []
        n = len(nums)
        result = []
        q = deque()
        for i in range(n):
            # Remove indices outside the window
            while q and q[0] < i - k + 1:
                q.popleft()
            # Remove smaller elements from the back
            while q and nums[q[-1]] < nums[i]:
                q.pop()
            # Add current index
            q.append(i)
            # Window is complete
            if i >= k - 1:
                result.append(nums[q[0]])
        return result

File: 0239-sliding-window-maximum/0239-sliding-window-maximum.cpp
class Solution {
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        if(nums.size()==0 || k<=0){
            return {} ;
        }
        int n = nums.size();
        vector<int>r(n-k+1);
        int ri=0;
        deque<int>q;
        for(int i =0;i<n;i++){
            //window ke bhr ane wale index ko remove karna hoga 
            while(!q.empty() && q.front()<i-k+1){
                q.pop_front();
            }
            //smaller elment ko rare se remove karna hoga 
            while(!q.empty() && nums[q.back()]<nums[i]){
                q.pop_back();

            }
            q.push_back(i);
        
        //jabh hamari window complete hogi tabh 
        if(i>=k-1){
            r[ri++]=nums[q.front()];
        }
    }
    return r;
    }
};